src/lib.rs

```diff
@@ -1308,6 +1308,35 @@ pub trait Itertools : Iterator {
         Ok(start)
     }
 
+    /// Fold using a mutable reference to the accumulator.
+    ///
+    /// Rather than having to mutate the accumulator and then return it,
+    /// this function allows simple, direct mutations.
+    ///
+    /// ```
+    /// use itertools::Itertools;
+    ///
+    /// #[derive(Debug, PartialEq)]
+    /// struct State(i32);
+    ///
+    /// impl State {
+    ///     fn mutate(&mut self, x: i32) {
+    ///         self.0 += x;
+    ///     }
+    /// }
+    ///
+    /// let s = (0..10).fold_mutref(State(0), |s_mut, x| s_mut.mutate(x));
+    /// assert_eq!(s, State(45));
+    /// ```
+    fn fold_mutref<ACC, F>(&mut self, start: ACC, mut f: F) -> ACC
+        where F: FnMut(&mut ACC, Self::Item)
+    {
+        self.fold(start, |mut acc, item| {
+            f(&mut acc, item);
+            acc
+        })
```

Comment:
I think a `foreach` (or `fold((), |(), elt| ..)`) based implementation is best, it's truest to the method that we don't move around the acc value at all, if that for some reason is a big inconvenience or does not optimize out.